tests/integration/test_xrdsst.py

```diff
@@ -1,25 +1,15 @@
-import json
-import os
-import subprocess
-import time
-import unittest
-from unittest import mock
-
-import docker
-import git
 import urllib3
 
-from definitions import ROOT_DIR
-from tests.util.test_util import get_service_description, find_test_ca_sign_url, perform_test_ca_sign, \
-    assert_server_statuses_transitioned
+from tests.integration.integration_base import IntegrationTestBase
+from tests.integration.integration_ops import IntegrationOpBase
 from tests.util.test_util import get_client, auth_cert_registration_global_configuration_update_received, waitfor
-from xrdsst.controllers.auto import AutoController
+from tests.util.test_util import get_service_description, assert_server_statuses_transitioned
```

Comment:
Integration tests fail:
FAILED tests/integration/test_xrdsst.py::TestXRDSST::test_run_configuration - UnboundLocalError: local variable 'user' referenced before assignment
FAILED tests/integration/test_xrdsst_auto.py::TestXrdsstAuto::test_run_auto_configuration - xrdsst.rest.rest.ApiException: (429)

Reply:
Rate limiting applied as well.

Reply:
Now it seems only the auto_configuration tests are not workiing:

=============================================================================================================== FAILURES ===============================================================================================================
______________________________________________________________________________________________ TestXrdsstAuto.test_run_auto_configuration ______________________________________________________________________________________________

self = <tests.integration.test_xrdsst_auto.TestXrdsstAuto testMethod=test_run_auto_configuration>

    def test_run_auto_configuration(self):
        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
        # Uninitialized security server can already have functioning API interface and status query must function
        unconfigured_servers_at_start = self.query_status()
    
        # Expected to run until the certificates that are not there
        self.step_autoconf()
    
        # Get signed certificates
        downloaded_csrs = self.step_cert_download_csrs()
        signed_certs = self.step_acquire_certs(downloaded_csrs)
    
        # Modify config to have certificates
        self.apply_cert_config(signed_certs)
    
        # Expected to import certificates, but not get further straight away, since registration globally is time-consuming.
        self.step_autoconf()
    
        # Wait for global configuration status updates
        waitfor(lambda: auth_cert_registration_global_configuration_update_received(self.config), self.retry_wait, self.max_retries)
        self.query_status()
    
        # Now that registered auth cert is globally accepted, should proceed with everything else to successful end.
        self.step_autoconf()
    
        configured_servers_at_end = self.query_status()
        assert_server_statuses_transitioned(unconfigured_servers_at_start, configured_servers_at_end)
    
        # Verify non-base operation transitions
>       assert_non_status_ops_transitioned(
            self.config["security_server"][0]["url"],
            self.config["security_server"][0]["api_key"]
        )

tests/integration/test_xrdsst_auto.py:102: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/integration/test_xrdsst_auto.py:66: in assert_non_status_ops_transitioned
    assert_clients_transitioned(api_url, api_key)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

api_url = 'https://172.17.0.2:4000/api/v1', api_key = 'X-Road-apikey token=7694dac1-c827-4586-9899-a6348b9fa474'

    def assert_clients_transitioned(api_url, api_key):
        clients = api_GET(api_url, "clients", api_key)
    
        assert (2 == len(clients))
>       assert (2 == len([x for x in clients if x['status'] == 'REGISTERED']))
E       AssertionError: assert 2 == 1
E         +2
E         -1

tests/integration/test_xrdsst_auto.py:13: AssertionError
--------------------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------------------

╒══════════════════╤════════════╤═══════════════════════╤═══════════════════════╤════════╤═════════╤════════╤════════╤═════════╕
│ GLOBAL           │ SERVER     │ ROLES                 │ INIT                  │ TSAS   │ TOKEN   │ KEYS   │ CSRS   │ CERTS   │
╞══════════════════╪════════════╪═══════════════════════╪═══════════════════════╪════════╪═════════╪════════╪════════╪═════════╡
│ FAIL (INTERNAL)  │ ss         │ System Administrator  │ TOKEN NOT_INITIALIZED │        │         │        │        │         │
│ LAST 134047 0226 │ VER 6.25.0 │ Service Administrator │                       │        │         │        │        │         │
│ NEXT 134147 0226 │            │ Registration Officer  │                       │        │         │        │        │         │
│                  │            │ Security Officer      │                       │        │         │        │        │         │
╘══════════════════╧════════════╧═══════════════════════╧═══════════════════════╧════════╧═════════╧════════╧════════╧═════════╛

AUTO ['init']->'ss'
Uploading configuration anchor for security server: ss
Upload of configuration anchor from "/home/bertvi/x/dev/test_toolkit/X-Road-Security-Server-toolkit/tests/resources/configuration-anchor.xml" successful
Initializing security server: ss
Security server "ss" initialized
AUTO ['timestamp init']->'ss'
ss

╒══════════╤═══════════════════════════╕
│ NAME     │ URL                       │
╞══════════╪═══════════════════════════╡
│ Test TSA │ http://10.249.34.187:8899 │
╘══════════╧═══════════════════════════╛

AUTO ['token login']->'ss'
Performing software token 0 login: 
Security server 'ss' token 0 logged in.
AUTO ['token init-keys']->'ss'
Generating software token 0 key labelled ss-default-auth-key and AUTH CSR: 
Generating software token 0 key labelled 'ss-default-sign-key' and SIGN CSR: 
Created AUTHENTICATION CSR '627C9570AB03680FADF66954AF4D667D496E6CC6' for key '075C1D50D137F0C1F6EAD0601F44D5E2B37D4070' as 'ss-default-auth-key'
Created SIGNING CSR '312B0CAE4B2BED536F97793B0AD2E70F0131853E' for key '3D3762C070517E757A63806EA87D77D54552D3E6' as 'ss-default-sign-key'
AUTO ['cert import']->'ss'
SKIPPED 'ss':
        'ss' [cert import]: 'certificates' missing required value. Submit downloaded CSRs for signing and specify received certificates accordingly.
AUTO ['cert import'] completion was NOT detected.
Some operations require waiting for global configuration renewal.
Next AUTO operation would have been ['cert register'].
AUTO ['status']->'ss' AT THE END OF AUTOCONFIGURATION.

╒══════════════════╤═════════════════╤═══════════════════════╤════════════════════╤══════════╤═════════════╤══════════╤══════════╤═════════╕
│ GLOBAL           │ SERVER          │ ROLES                 │ INIT               │ TSAS     │ TOKEN       │ KEYS     │ CSRS     │ CERTS   │
╞══════════════════╪═════════════════╪═══════════════════════╪════════════════════╪══════════╪═════════════╪══════════╪══════════╪═════════╡
│ FAIL (INTERNAL)  │ ss              │ System Administrator  │ ANCHOR INITIALIZED │ Test TSA │ ID 0        │ SIGN (1) │ SIGN (1) │         │
│ LAST 134047 0226 │ VER 6.25.0      │ Service Administrator │ CODE INITIALIZED   │          │ softToken-0 │ AUTH (1) │ AUTH (1) │         │
│ NEXT 134147 0226 │ DEV:GOV:1234:SS │ Registration Officer  │ OWNER INITIALIZED  │          │ STATUS OK   │ 2 KEYS   │ 2 CSRS   │         │
│                  │                 │ Security Officer      │ TOKEN INITIALIZED  │          │ LOGIN YES   │          │          │         │
╘══════════════════╧═════════════════╧═══════════════════════╧════════════════════╧══════════╧═════════════╧══════════╧══════════╧═════════╛

Starting CSR download from security server: ss

╒══════════════════════════════════════════╤══════════════════════════════════════════╤════════╤════════════════════════════════════════════════════════════════════════╕
│ CSR ID                                   │ KEY ID                                   │ TYPE   │ LOCATION                                                               │
╞══════════════════════════════════════════╪══════════════════════════════════════════╪════════╪════════════════════════════════════════════════════════════════════════╡
│ 312B0CAE4B2BED536F97793B0AD2E70F0131853E │ 3D3762C070517E757A63806EA87D77D54552D3E6 │ SIGN   │ /tmp/ss-sign-CSR-312B0CAE4B2BED536F97793B0AD2E70F0131853E-4ril8qsy.der │
├──────────────────────────────────────────┼──────────────────────────────────────────┼────────┼────────────────────────────────────────────────────────────────────────┤
│ 627C9570AB03680FADF66954AF4D667D496E6CC6 │ 075C1D50D137F0C1F6EAD0601F44D5E2B37D4070 │ AUTH   │ /tmp/ss-auth-CSR-627C9570AB03680FADF66954AF4D667D496E6CC6-nmimoeg9.der │
╘══════════════════════════════════════════╧══════════════════════════════════════════╧════════╧════════════════════════════════════════════════════════════════════════╛

AUTO ['init']->'ss' (redo) 
Configuration anchor for "ss" already imported
Security server "ss" already initialized
AUTO ['timestamp init']->'ss' (redo) 
ss Timestamping service already configured.
AUTO ['token login']->'ss' (redo) 
Performing software token 0 login: 
Token 0 already logged in for 'ss'.
AUTO ['token init-keys']->'ss' (redo) 
No key initialization needed.
AUTO ['cert import']->'ss'
Imported certificate '/tmp/ss-sign-CSR-312B0CAE4B2BED536F97793B0AD2E70F0131853E-4ril8qsy.der.signed.pem'
Imported certificate '/tmp/ss-auth-CSR-627C9570AB03680FADF66954AF4D667D496E6CC6-nmimoeg9.der.signed.pem'
AUTO ['cert register']->'ss'
Registered certificate 7E190304CDD3097685917272FFC3987C8382F589 for address '{'address': '172.17.0.2'}'
AUTO ['cert activate']->'ss'
Activated certificate 7E190304CDD3097685917272FFC3987C8382F589
AUTO ['client add']->'ss' (redo) 
Added client subsystem GOV:1234:BUS (got full id DEV:GOV:1234:BUS)
AUTO ['service add-description']->'ss' (redo) 
Added service description with type 'OPENAPI3' and url 'https://raw.githubusercontent.com/OpenAPITools/openapi-generator/master/modules/openapi-generator-gradle-plugin/samples/local-spec/petstore-v3.0.yaml' (got full id 8)
AUTO ['service enable-description']->'ss' (redo) 
Service description for 'GOV:1234:BUS' with id: '8' enabled successfully.
AUTO ['client register']->'ss' (redo) 
AUTO ['status']->'ss' AT THE END OF AUTOCONFIGURATION.

╒══════════════════╤═════════════════╤═══════════════════════╤════════════════════╤══════════╤═════════════╤══════════╤════════╤═════════╕
│ GLOBAL           │ SERVER          │ ROLES                 │ INIT               │ TSAS     │ TOKEN       │ KEYS     │ CSRS   │ CERTS   │
╞══════════════════╪═════════════════╪═══════════════════════╪════════════════════╪══════════╪═════════════╪══════════╪════════╪═════════╡
│ FAIL (INTERNAL)  │ ss              │ System Administrator  │ ANCHOR INITIALIZED │ Test TSA │ ID 0        │ SIGN (1) │        │ SIGN    │
│ LAST 134047 0226 │ VER 6.25.0      │ Service Administrator │ CODE INITIALIZED   │          │ softToken-0 │ AUTH (1) │        │ AUTH    │
│ NEXT 134147 0226 │ DEV:GOV:1234:SS │ Registration Officer  │ OWNER INITIALIZED  │          │ STATUS OK   │ 2 KEYS   │        │         │
│                  │                 │ Security Officer      │ TOKEN INITIALIZED  │          │ LOGIN YES   │          │        │         │
╘══════════════════╧═════════════════╧═══════════════════════╧════════════════════╧══════════╧═════════════╧══════════╧════════╧═════════╛


╒══════════════════╤═════════════════╤═══════════════════════╤════════════════════╤══════════╤═════════════╤══════════╤════════╤═════════╕
│ GLOBAL           │ SERVER          │ ROLES                 │ INIT               │ TSAS     │ TOKEN       │ KEYS     │ CSRS   │ CERTS   │
╞══════════════════╪═════════════════╪═══════════════════════╪════════════════════╪══════════╪═════════════╪══════════╪════════╪═════════╡
│ OK (SUCCESS)     │ ss              │ System Administrator  │ ANCHOR INITIALIZED │ Test TSA │ ID 0        │ SIGN (1) │        │ SIGN    │
│ LAST 134347 0226 │ VER 6.25.0      │ Service Administrator │ CODE INITIALIZED   │          │ softToken-0 │ AUTH (1) │        │ AUTH    │
│ NEXT 134447 0226 │ DEV:GOV:1234:SS │ Registration Officer  │ OWNER INITIALIZED  │          │ STATUS OK   │ 2 KEYS   │        │         │
│                  │                 │ Security Officer      │ TOKEN INITIALIZED  │          │ LOGIN YES   │          │        │         │
╘══════════════════╧═════════════════╧═══════════════════════╧════════════════════╧══════════╧═════════════╧══════════╧════════╧═════════╛

AUTO ['init']->'ss' (redo) 
Configuration anchor for "ss" already imported
Security server "ss" already initialized
AUTO ['timestamp init']->'ss' (redo) 
ss Timestamping service already configured.
AUTO ['token login']->'ss' (redo) 
Performing software token 0 login: 
Token 0 already logged in for 'ss'.
AUTO ['token init-keys']->'ss' (redo) 
No key initialization needed.
AUTO ['cert import']->'ss' (redo) 
Certificate '/tmp/ss-sign-CSR-312B0CAE4B2BED536F97793B0AD2E70F0131853E-4ril8qsy.der.signed.pem' already imported.
Certificate '/tmp/ss-auth-CSR-627C9570AB03680FADF66954AF4D667D496E6CC6-nmimoeg9.der.signed.pem' already imported.
AUTO ['cert register']->'ss' (redo) 
No certificates to 'REGISTER' for key labelled 'ss-default-auth-key'.
AUTO ['cert activate']->'ss' (redo) 
No certificates to 'ACTIVATE' for key labelled 'ss-default-auth-key'.
AUTO ['client add']->'ss' (redo) 
Client for 'GOV:1234:BUS' already exists.
AUTO ['service add-description']->'ss' (redo) 
Service description for 'GOV:1234:BUS' with url 'https://raw.githubusercontent.com/OpenAPITools/openapi-generator/master/modules/openapi-generator-gradle-plugin/samples/local-spec/petstore-v3.0.yaml' and type 'OPENAPI3' already exists.
AUTO ['service enable-description']->'ss' (redo) 
Service description for 'GOV:1234:BUS' with id: '8' enabled successfully.
AUTO ['client register']->'ss' (redo) 
AUTO ['status']->'ss' AT THE END OF AUTOCONFIGURATION.

╒══════════════════╤═════════════════╤═══════════════════════╤════════════════════╤══════════╤═════════════╤══════════╤════════╤═════════╕
│ GLOBAL           │ SERVER          │ ROLES                 │ INIT               │ TSAS     │ TOKEN       │ KEYS     │ CSRS   │ CERTS   │
╞══════════════════╪═════════════════╪═══════════════════════╪════════════════════╪══════════╪═════════════╪══════════╪════════╪═════════╡
│ OK (SUCCESS)     │ ss              │ System Administrator  │ ANCHOR INITIALIZED │ Test TSA │ ID 0        │ SIGN (1) │        │ SIGN    │
│ LAST 134347 0226 │ VER 6.25.0      │ Service Administrator │ CODE INITIALIZED   │          │ softToken-0 │ AUTH (1) │        │ AUTH    │
│ NEXT 134447 0226 │ DEV:GOV:1234:SS │ Registration Officer  │ OWNER INITIALIZED  │          │ STATUS OK   │ 2 KEYS   │        │         │
│                  │                 │ Security Officer      │ TOKEN INITIALIZED  │          │ LOGIN YES   │          │        │         │
╘══════════════════╧═════════════════╧═══════════════════════╧════════════════════╧══════════╧═════════════╧══════════╧════════╧═════════╛


╒══════════════════╤═════════════════╤═══════════════════════╤════════════════════╤══════════╤═════════════╤══════════╤════════╤═════════╕
│ GLOBAL           │ SERVER          │ ROLES                 │ INIT               │ TSAS     │ TOKEN       │ KEYS     │ CSRS   │ CERTS   │
╞══════════════════╪═════════════════╪═══════════════════════╪════════════════════╪══════════╪═════════════╪══════════╪════════╪═════════╡
│ OK (SUCCESS)     │ ss              │ System Administrator  │ ANCHOR INITIALIZED │ Test TSA │ ID 0        │ SIGN (1) │        │ SIGN    │
│ LAST 134347 0226 │ VER 6.25.0      │ Service Administrator │ CODE INITIALIZED   │          │ softToken-0 │ AUTH (1) │        │ AUTH    │
│ NEXT 134447 0226 │ DEV:GOV:1234:SS │ Registration Officer  │ OWNER INITIALIZED  │          │ STATUS OK   │ 2 KEYS   │        │         │
│                  │                 │ Security Officer      │ TOKEN INITIALIZED  │          │ LOGIN YES   │          │        │         │
╘══════════════════╧═════════════════╧═══════════════════════╧════════════════════╧══════════╧═════════════╧══════════╧════════╧═════════╛

ss
--------------------------------------------------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------------------------------------------------
Exception calling ClientsApi->register_client: (500)
Reason: 
HTTP response headers: HTTPHeaderDict({'x-road-ui-correlation-id': '4d8f69994e039796', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': '0', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'none'", 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Date': 'Fri, 26 Feb 2021 13:41:35 GMT', 'Connection': 'close'})
HTTP response body: {"status":500,"error":{"code":"core.Server.ClientProxy.SslAuthenticationFailed","metadata":["Security server has no valid authentication certificate"]}}

Exception calling ClientsApi->register_client: (500)
Reason: 
HTTP response headers: HTTPHeaderDict({'x-road-ui-correlation-id': '8958d8891cf11572', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': '0', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'none'", 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Date': 'Fri, 26 Feb 2021 13:44:18 GMT', 'Connection': 'close'})
HTTP response body: {"status":500,"error":{"code":"core.Server.ClientProxy.NetworkError","metadata":["SS1: No address associated with hostname"]}}

---------------------------------------------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------------------------------------------
ERROR    root:base.py:409 Exception calling ClientsApi->register_client: (500)
Reason: 
HTTP response headers: HTTPHeaderDict({'x-road-ui-correlation-id': '4d8f69994e039796', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': '0', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'none'", 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Date': 'Fri, 26 Feb 2021 13:41:35 GMT', 'Connection': 'close'})
HTTP response body: {"status":500,"error":{"code":"core.Server.ClientProxy.SslAuthenticationFailed","metadata":["Security server has no valid authentication certificate"]}}

ERROR    root:base.py:409 Exception calling ClientsApi->register_client: (500)
Reason: 
HTTP response headers: HTTPHeaderDict({'x-road-ui-correlation-id': '8958d8891cf11572', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': '0', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'none'", 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Date': 'Fri, 26 Feb 2021 13:44:18 GMT', 'Connection': 'close'})
HTTP response body: {"status":500,"error":{"code":"core.Server.ClientProxy.NetworkError","metadata":["SS1: No address associated with hostname"]}}

----------- coverage: platform linux, python 3.8.5-final-0 -----------
Name                              Stmts   Miss  Cover
-----------------------------------------------------
xrdsst/controllers/auto.py           58      3    95%
xrdsst/controllers/base.py          299     13    96%
xrdsst/controllers/cert.py          190      8    96%
xrdsst/controllers/client.py         91      5    95%
xrdsst/controllers/init.py           62      0   100%
xrdsst/controllers/service.py       102      5    95%
xrdsst/controllers/status.py         53      3    94%
xrdsst/controllers/timestamp.py     101      7    93%
xrdsst/controllers/token.py         163     10    94%
-----------------------------------------------------
TOTAL                              1119     54    95%
Coverage HTML written to dir coverage-report

======================================================================================================= short test summary info ========================================================================================================
FAILED tests/integration/test_xrdsst_auto.py::TestXrdsstAuto::test_run_auto_configuration - AssertionError: assert 2 == 1

Reply:
Does not reproduce, 
```
'core.Server.ClientProxy.SslAuthenticationFailed' & "code":"core.Server.ClientProxy.NetworkError","metadata":["SS1: No address associated with hostname"]} 
```
is probably  either mark of rate limit still being exceeded in tests or bad connectivity between central server and management server.